### PR TITLE
kam1a  adding convert method to collection (closes #917)

### LINKED ISSUE #917: collection.convert
Would like to support syntax `collection.convert(unit)`.

Would convert all contained data objects recursively.

### PR DESCRIPTION
**Pull request by kameyer226**
## Changes

adding convert method to collection (closes #917)

## Checklist

- [x] added tests, if applicable
- [x] updated documentation, if applicable
- [x] updated CHANGELOG.md
- [x] tests pass

### CODE REVIEW

**Comment by kameyer226** [comment]
Weirdness with git I have to puzzle out, resolved by making changes to a new branch instead.

**Comment by untzag** [comment]
@kameyer226 please try to keep a single change within a single branch. You can commit many times to the same branch and the same PR. Feel free to reach out or look online if you are having problems with git. Remember---read the messages carefully and git often tells you exactly what to do :smile: 

When you open up many branches / PRs all about the same issue, it becomes challenging to see exactly what you're working on as a maintainer.

This resource might help: https://guides.github.com/introduction/flow/

**Comment by kameyer226** [comment]
I did the normal route, but I got an error saying the remote contains work I did not have locally.  Thinking someone modified the remote, I re pulled.    It then asked me which branch to merge with (917_KAM1 or pint), which didn't make sense to me, as why should my remote origin need to be respec'd after something so mundane as a pull off the branch I was working on?

**Comment by kameyer226** [comment]
PS C:\work\wright\python\WrightTools\WrightTools> git commit -am "WIP_917_KAM1a"
[917_KAM1 b1c8ba6] WIP_917_KAM1a
PS C:\work\wright\python\WrightTools\WrightTools> git push
To https://github.com/wright-group/WrightTools
 ! [rejected]        917_KAM1 -> 917_KAM1 (fetch first)
error: failed to push some refs to 'https://github.com/wright-group/WrightTools'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
PS C:\work\wright\python\WrightTools\WrightTools> git pull
remote: Enumerating objects: 54, done.
remote: Counting objects: 100% (54/54), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 54 (delta 38), reused 48 (delta 32), pack-reused 0
Unpacking objects: 100% (54/54), 10.90 KiB | 36.00 KiB/s, done.
From https://github.com/wright-group/WrightTools
   0a27d84..efdc3f1  917_KAM1   -> origin/917_KAM1
 + d870114...f43731f pint       -> origin/pint  (forced update)
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> 917_KAM1

PS C:\work\wright\python\WrightTools\WrightTools>

**Comment by untzag** [comment]
Yep, just do what git is asking:

```
git branch --set-upstream-to=origin/917_KAM1 917_KAM1
```

File: WrightTools/collection/_collection.py [hunk 1]
```diff
@@ -102,6 +102,17 @@ def create_collection(self, name="collection", position=None, **kwargs):
...
 
+    def convert(self, units):
+        """Convert units of a collection.
+        Parameters
+        ----------
+        units: string
+            Units to convert to.
+        """
+
+        for data in self.data:
```
Comment (ksunden) [inline WrightTools/collection/_collection.py:113]: Collection objects have no attribute `data`, I believe you can just do `for i in self:` or `for i in self.items():`

**Comment by ksunden** [comment]
I would also suggest adding at least the `verbose` kwarg, maybe even `convert_variables` (each just passed into the underlying calls)

This will keep the api consistent, by providing the same options.

https://github.com/wright-group/WrightTools/blob/123a9b1c922e83c5e1c3b39c39279aea134e80ce/WrightTools/data/_data.py#L761

**Comment by untzag** [comment]
@kameyer226 could you please update the title and description of this PR? Thanks :smile:

**Comment by ksunden** [comment]
@kameyer226 Can you please put a descriptive title and description in this PR (you can edit with the button on the top right)

That will help others to know what is going on without needing to go searching for the issue.

You should put something like "Closes #917" in the description, this will link the PR to the issue so that the issue will be automatically closed upon merge of the PR.

Additionally, I highly recommend having descriptive commit messages:

e.g.:
`git commit -am "remove infinite recursion"`

(note that the quotes allow you to have spaces, even new lines)

**Comment by ksunden** [comment]
Note that it only gets linked properly if the `Closes #917` is in the _description_ (not the title... its fine for it to be there as well, just doesn't do the auto close)

**Comment by kameyer226** [comment]
Perhaps I am not certain what units are to be converted.   Calling `Collection()` currently does not instantiate a `Data` object, however, by calling `Collection.create_data()` a Data object becomes instantiated.   I thought one only needed to convert the units of that Data object.

**Comment by ksunden** [comment]
A collection is is a folder of potentially many data objects.

when you do `self.data` you are only getting one data object, and only then if it is itself called "data" (the default for `create_data`, but by no means guaranteed to exist)

This method should loop over _all_ data objects (and indeed nested collections, which are also allowed), no matter their names.

**Comment by ksunden** [comment]
http://wright.tools/en/stable/collection.html

**Comment by kameyer226** [comment]
I see, I had not come across an example where a collection was nested inside a collection.    Does this nesting continue beyond one level?

**Comment by ksunden** [comment]
It can, but that does not matter in the slightest to the implementation.

**Comment by ksunden** [comment]
You do not need to do _any_ isinstance check

File: WrightTools/collection/_collection.py [hunk 2]
```diff
@@ -102,6 +102,25 @@ def create_collection(self, name="collection", position=None, **kwargs):
...
+        for name in self.item_names:
+            item = self[name]
+            if isinstance(item, wt_data.Data):
+                item.convert(units, convert_variables=convert_variables, verbose=verbose)
+            elif isinstance(item, Collection):
+                Collection.convert(
+                    item, units, convert_variables=convert_variables, verbose=verbose
+                )
+            else:
+                pass
```
Comment (untzag) [inline WrightTools/collection/_collection.py:122]: @kameyer226 consider removing the isinstance checks. Collections and Data objects have the same function signature for `convert`, so you should be able to just blindly call without actually knowing what kind of item you have. I think so, at least :thinking: 

```suggestion
            item.convert(units, convert_variables=convert_variables, verbose=verbose)
```

**Comment by kameyer226** [comment]
Results of script run:   (Forgot to ask how valid test scripts should be reported to reviewers)

axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
axis w1 converted from wn to eV
axis w2 converted from wn to eV
True
True
True
True
True
True
True
True

**Comment by kameyer226** [comment]
Sorry to respond late to @untzag :  I left the isinstances separate because I wanted to create the impression that the collection may consist of objects that are not type Data

**Comment by kameyer226** [comment]
Somehow .vscode is not in the exceptions to git, guess I better delete...

**Review by untzag — APPROVED** [review]
:tada:

**Pull request merged**